### PR TITLE
[EC-63] Implement breadcrumb component

### PR DESCRIPTION
**Pull request by coroiu**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Code changes

- **file.ext:** Description of what was changed and why

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Comment by coroiu** [comment]
@bitwarden/dept-design should the ellipsis use link button or a small icon button?

Link button (hover has no visible changes):
image

Icon button:
image

image

**Comment by danielleflinn** [comment]
@coroiu the small icon button seems good since there is a visible hover state

**Comment by coroiu** [comment]
> @coroiu the small icon button seems good since there is a visible hover state. Did you have any concerns about going that route?

@danielleflinn Not really I was just unsure :) The padding is going to be a bit bigger than on the design in figma (which the link button can replicate a bit better)

**Comment by danielleflinn** [comment]
Padding looks good to me in the screenshot; thanks for calling out the difference though.

File: libs/common/src/models/domain/tree-node.ts [hunk 1]
```diff
@@ -1,9 +1,9 @@
 export class TreeNode<T extends ITreeNodeObject> {
-  parent: T;
   node: T;
+  parent: TreeNode<T>;
   children: TreeNode<T>[] = [];
 
-  constructor(node: T, parent: T, name?: string, id?: string) {
+  constructor(node: T, parent: TreeNode<T>, name?: string, id?: string) {
```
Comment (jlf0dev) [inline libs/common/src/models/domain/tree-node.ts:6]: Nicely done, should've been a `TreeNode<T>` all along

File: libs/common/src/misc/serviceUtils.spec.ts [hunk 2]
```diff
@@ -82,3 +53,11 @@ describe("serviceUtils", () => {
     });
   });
 });
+
+const createTreeNode =
+  (obj: FakeObject, children: ((parent: TreeNode<FakeObject>) => TreeNode<FakeObject>)[] = []) =>
+  (parent?: TreeNode<FakeObject>): TreeNode<FakeObject> => {
```
Comment (jlf0dev) [inline libs/common/src/misc/serviceUtils.spec.ts:59]: Cool, pretty hard to read this signature though.

File: apps/web/src/app/vault/vault.component.html [hunk 3]
```diff
@@ -17,6 +17,18 @@
...
+      <bit-breadcrumbs *ngIf="breadcrumbs.length > 0">
+        <bit-breadcrumb
+          *ngFor="let collection of breadcrumbs; let first = first"
+          [icon]="first ? undefined : 'bwi-collection'"
+          (click)="applyCollectionFilter(collection)"
+        >
+          <ng-container *ngIf="first; else notTranslated">{{
+            collection.node.name | i18n
+          }}</ng-container>
+          <ng-template #notTranslated>{{ collection.node.name }}</ng-template>
```
Comment (jlf0dev) [inline apps/web/src/app/vault/vault.component.html:29]: `#notTranslated` also being used here

**Review by jlf0dev — APPROVED** [review]
Everything looks good, just had a nitpick and some comments. Going to approve for now, let me know if you make any changes and need another review!

File: libs/common/src/misc/serviceUtils.spec.ts (hunk 2 above)
Comment (coroiu) [reply libs/common/src/misc/serviceUtils.spec.ts:59]: Hmm good point, not sure exactly how to improve it though 🤔

File: apps/web/src/app/organizations/vault/vault.component.html [hunk 4]
```diff
@@ -16,6 +16,18 @@
       </div>
     </div>
     <div class="col-9">
+      <bit-breadcrumbs *ngIf="breadcrumbs.length > 0">
+        <bit-breadcrumb
+          *ngFor="let collection of breadcrumbs; let first = first"
+          [icon]="first ? undefined : 'bwi-collection'"
+          (click)="applyCollectionFilter(collection)"
+        >
+          <ng-container *ngIf="first; else notTranslated">{{
```
[the remark this answers is not in the document]
Comment (coroiu) [reply apps/web/src/app/organizations/vault/vault.component.html:25]: Oh yeah, actually this is code from like months ago before we had the discussion regarding `else` vs double `ngIf`, I'll fix!

File: apps/web/src/app/shared/components/breadcrumbs/breadcrumb.component.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+})
+export class BreadcrumbComponent {
+  @Input()
+  icon?: string;
+
+  @Input()
+  route?: string | any[] = undefined;
+
+  @Input()
+  queryParams?: Record<string, string> = {};
```
[the remark this answers is not in the document]
Comment (coroiu) [reply apps/web/src/app/shared/components/breadcrumbs/breadcrumb.component.ts:15]: Hmmm, yeah but at the same time I'm not sure that the breadcrumb component would be able to use it

**Comment by coroiu** [comment]
@jlf0dev apparently your approval wasn't dismissed, but I'd feel better if you took another look!

**Comment by jlf0dev** [comment]
LGTM!

**Pull request merged**